### PR TITLE
[Orphaned] perlin noise generator

### PR DESCRIPTION
**Pull request by zuntrax**
C++ rewrite in progress.
* [x] C++ rewrite of Perlin class
* [ ] documentation
* [ ] test case or demo for Perlin class
* [x] delete python implementation

Should be useful for #65.

These were generated with the python implementation, the new ones will look different.
Seed: "sftworld1", 500x500px, 10x10 grid cells, polynomial interpolation:
perlin_sftworld1

Red: "sftworld0", green: "sftworld1", blue: "sftworld2", polynomial interpolation, dims as above:
multicolor

### CODE REVIEW

**Comment by franciscod** [comment]
awesome! :+1:

**Comment by rkreis** [comment]
Just created a 10x10 random image and scaled it:
interp

I'm taking a closer look at `perlin_mdim_inter`.

**Comment by zuntrax** [comment]
@rkreis I'll add some debug output and check if it works how I think it should work. I'll poke you.

**Comment by TheJJ** [comment]
update to the new project layout please.

**Comment by zuntrax** [comment]
Will update asap.

**Comment by franciscod** [comment]
is this still relevant? i mean we already have some kind of map generation, i don't know if this branch is complementary to that

**Comment by TheJJ** [comment]
The current map generation isn't capable of generating epic minecraft-like terrain, which is only possible with intended functionality of this pull. @zuntrax comeon, work on it again!

**Comment by zuntrax** [comment]
Ohai, I'm still alive. I just realized that this pull is more than a year old. Time to get this working.

File: libopenage/perlin/perlin.h [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+	seed_t seed;
+	size_t granularity;
+	std::unordered_map<cvec_t, vvec_t> gradient_cache;
+	// Maximum possible gradient vector component value without risking
+	// overflows during dot_product calculations.
+	value_t comp_max = static_cast<value_t>(sqrt(static_cast<double>(INT64_MAX) / N));
+
+	value_t interpolate(value_t x, value_t y, double pos) const {
+		double factor = (3 * std::pow(pos, 2)) - (2 * std::pow(pos, 3));
```
Comment (fr3akout) [inline libopenage/perlin/perlin.h]: std::pow(pos, 2) == pos * pos

File: libopenage/perlin/perlin.h [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+		auto res = vvec_t{};
+		for (size_t i = 0; i < N; i++) {
+			res[i] = static_cast<value_t>(this->comp_max * tmp[i]);
+		}
+		return res;
+	}
+
+	vvec_t get_gradient(const cvec_t &node) {
+		auto it = this->gradient_cache.find(node);
+		if (it != this->gradient_cache.end()) {
```
Comment (fr3akout) [inline libopenage/perlin/perlin.h]: c++11-style: std::end(this->gradient_cache)

File: libopenage/perlin/perlin.h [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+		// backup relevant state of point
+		coord_t saved_coord = point[dim];
+
+		point[dim] = this->lower_cell_border(saved_coord);
+		value_t lo_part = this->multidim_inter(dim + 1, point, orig_point);
+
+		point[dim] += this->granularity;
+		value_t hi_part = this->multidim_inter(dim + 1, point, orig_point);
+
+		point[dim] = saved_coord;
```
Comment (fr3akout) [inline libopenage/perlin/perlin.h]: Is this really a noticable performance drop without optimization?

File: libopenage/util/generator.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+#include <cassert>
+
+namespace openage {
+namespace util {
+
+/**
+ * Abstract class for imitating python generators.
+ */
+template<class T>
+class Generator {
```
Comment (fr3akout) [inline libopenage/util/generator.h]: missing virtual destructor

File: libopenage/perlin/perlin.h [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+/**
+ */
+template<size_t N>
+class Perlin {
+	using cvec_t = coord_vec<N>;
+	using vvec_t = value_vec<N>;
+
+	seed_t seed;
+	size_t granularity;
+	std::unordered_map<cvec_t, vvec_t> gradient_cache;
```
Comment (fr3akout) [inline libopenage/perlin/perlin.h]: Maybe extract generalized cache class?

**Comment by johanhelsing** [comment]
@zuntrax Any reason you chose Perlin over Simplex? Is that what's used in the original game?

**Comment by zuntrax** [comment]
@johanhelsing Perlin has the advantage that 2d slices through 3d noise look like 2d noise. This means that we can generate similar, gradually changing maps. I have no idea if this will actually be used.
Simplex noise is particularly interesting in 4 and 5 dimensions, because of its better time complexity. However, simplex yields garbage values in 6 or more dimensions. Also it is under software patents in the US. We could safely ignore this, since we are from Germany, but it is still a reason to dislike it.

**Comment by zuntrax** [comment]
much integer. such overflow. wow
noise

**Comment by zuntrax** [comment]
I won't be working on this any time soon, if someone wants, feel free to continue...

**Pull request closed without merging**